frontend/app/controllers/PayPal.scala

```diff
@@ -22,22 +25,25 @@ object PayPal extends Controller with LazyLogging {
 
 	// Sets up a payment by contacting PayPal, returns the token as JSON.
 	def setupPayment = NoCacheAction { request =>
-    logger.info("Called setupPayment")
-		Ok(tokenJsonResponse(PayPalService.retrieveToken(request)))
+    parseJsonAndRunServiceCall(request)(PayPalService.retrieveToken(request))
 	}
 
 	// Creates a billing agreement using a payment token.
 	def createAgreement = NoCacheAction { request =>
-		request.body.asJson.map { json =>
+    parseJsonAndRunServiceCall(request)(PayPalService.retrieveBaid)
+	}
 
-			Json.fromJson[Token](json) match {
-				case JsSuccess(token: Token, _) => Ok(tokenJsonResponse(PayPalService.retrieveBaid(token)))
-				case e: JsError => BadRequest(JsError.toJson(e).toString)
-			}
+  //Takes a request, parses it into a type T, passes this into serviceCall to retrieve a token then returns this as json
+  def parseJsonAndRunServiceCall[T](request:Request[AnyContent])(serviceCall : T => String)(implicit fjs: Reads[T]) = {
+    request.body.asJson.map { json =>
 
-		}.getOrElse(BadRequest)
+      Json.fromJson[T](json)(fjs) match {
+        case JsSuccess(parsed, _) => Ok(tokenJsonResponse(serviceCall(parsed)))
+        case e: JsError => BadRequest(JsError.toJson(e).toString)
```

Comment:
Yes I think it's just for debugging but as discussed it is only showing parsing errors from the request body which has come from our page and so is public anyway.
I'm going to leave it as is because it can be useful for debugging during development